### PR TITLE
Parse kind name passed in ENV var and use it when creating tieried kind

### PR DESCRIPTION
**Pull request by bratpiorka**
### Types of changes

- [ ] Bugfix (non-breaking change which fixes issue linked in Description above)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (correction or otherwise)
- [ ] Cosmetics (whitespace, appearance)
- [ ] Other

### Checklist

- [ ] Code compiles without errors
- [ ] All tests pass locally with my changes (see TESTING section in CONTRIBUTING file)
- [ ] Created tests which will fail without the change (if possible)
- [ ] Extended the README/documentation (if necessary)
- [ ] All newly added files have proprietary license (if necessary)
- [ ] All newly added files are referenced in MANIFEST files (if necessary)

## Further comments

---
This change is [Reviewable](https://reviewable.io/reviews/memkind/memkind/570)

### CODE REVIEW

**Review by bratpiorka — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/570)* status: 0 of 4 files reviewed, 2 unresolved discussions (waiting on @michalbiesek and @PatKamin)

___
*[tiering/ctl.c, line 46 at r1](https://reviewable.io/reviews/memkind/memkind/570#-MVqR45A7RfrZ2bOVHee:-MVqR45BEu1wskE-8qOq:b-e6cdda) ([raw file](https://github.com/memkind/memkind/blob/51a3ea47cc053d53fdd3ffd5ecd383d7b26ac526/tiering/ctl.c#L46)):*
> ```C
>         *kind = MEMKIND_DEFAULT;
>     } else if (strcmp(kind_name, "FS_DAX") == 0) {
>         *kind = MEMKIND_DAX_KMEM;
> ```

here I'm not sure if FS_DAX == MEMKIND_DAX_KMEM 
___
*[tiering/memtier.c, line 5 at r1](https://reviewable.io/reviews/memkind/memkind/570#-MVqRAl1E7WIT3uepGzA:-MVqRAl1E7WIT3uepGzB:b-oeucwl) ([raw file](https://github.com/memkind/memkind/blob/51a3ea47cc053d53fdd3ffd5ecd383d7b26ac526/tiering/memtier.c#L5)):*
> ```C
> 
> #include "../config.h"
> #include <memkind/internal/memkind_private.h>
> ```

needed for extracting kind->name - this could be removed when we remove obsolete logging

**Review by michalbiesek — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/570)* status: 0 of 4 files reviewed, 3 unresolved discussions (waiting on @bratpiorka and @PatKamin)

___
*[tiering/ctl.c, line 46 at r1](https://reviewable.io/reviews/memkind/memkind/570#-MVqR45A7RfrZ2bOVHee:-MVqTL8d1Kn0KFyxGb6-:b4ti6mb) ([raw file](https://github.com/memkind/memkind/blob/51a3ea47cc053d53fdd3ffd5ecd383d7b26ac526/tiering/ctl.c#L46)):*
> ```
> memkind_dax_kmem
> ```
FS_DAX is kind created using memkind_create_pmem API the MEMKIND_DAX_KMEM is a completely different thing.
___
*[tiering/memtier.c, line 5 at r1](https://reviewable.io/reviews/memkind/memkind/570#-MVqRAl1E7WIT3uepGzA:-MVqTXEj4BRbf_lY5jze:b-6pabi2) ([raw file](https://github.com/memkind/memkind/blob/51a3ea47cc053d53fdd3ffd5ecd383d7b26ac526/tiering/memtier.c#L5)):*
<details><summary><i>Previously, bratpiorka (Rafał Rudnicki) wrote…</i></summary><blockquote>

needed for extracting kind->name - this could be removed when we remove obsolete logging
</blockquote></details>

Yes this needs to be removed mark this as TODO
___
*[tiering/memtier.c, line 119 at r1](https://reviewable.io/reviews/memkind/memkind/570#-MVqToU_B-wswN-Pryiv:-MVqToU_B-wswN-Pryiw:bo1bfdz) ([raw file](https://github.com/memkind/memkind/blob/51a3ea47cc053d53fdd3ffd5ecd383d7b26ac526/tiering/memtier.c#L119)):*
> ```C
>     log_debug("pmem_size: %s", pmem_size);
>     log_debug("ratio_value: %u", ratio_value);
> ```

IMHO: This should be a place to handle kind creation/ memkind_t setup, ctl module should be responsible for parsing string.

**Review by PatKamin — COMMENTED** [review]
> Reviewed 4 of 4 files at r1.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/570)* status: all files reviewed, 4 unresolved discussions (waiting on @bratpiorka)

___
*[test/tiering_tests.py, line 198 at r1](https://reviewable.io/reviews/memkind/memkind/570#-MVtiz4Z46HjI7e0qSUh:-MVtiz4Z46HjI7e0qSUi:b1az8ez) ([raw file](https://github.com/memkind/memkind/blob/51a3ea47cc053d53fdd3ffd5ecd383d7b26ac526/test/tiering_tests.py#L198)):*
> ```Python
>     # TODO enable this check after implementing full FS_DAX support in
>     # libmemtier
>     """
> ```

Comment out just the kind_name check.
___
*[tiering/ctl.c, line 46 at r1](https://reviewable.io/reviews/memkind/memkind/570#-MVqR45A7RfrZ2bOVHee:-MVtkB0d9lzdghyo2e9R:b32v57) ([raw file](https://github.com/memkind/memkind/blob/51a3ea47cc053d53fdd3ffd5ecd383d7b26ac526/tiering/ctl.c#L46)):*
<details><summary><i>Previously, michalbiesek wrote…</i></summary><blockquote>

> ```
> memkind_dax_kmem
> ```
FS_DAX is kind created using memkind_create_pmem API the MEMKIND_DAX_KMEM is a completely different thing.
</blockquote></details>

+1

**Review by bratpiorka — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/570)* status: 0 of 2 files reviewed, 2 unresolved discussions (waiting on @michalbiesek and @PatKamin)

___
*[test/tiering_tests.py, line 198 at r1](https://reviewable.io/reviews/memkind/memkind/570#-MVtiz4Z46HjI7e0qSUh:-MVu-W7hDknaqY2R-aZC:bth9l0b) ([raw file](https://github.com/memkind/memkind/blob/51a3ea47cc053d53fdd3ffd5ecd383d7b26ac526/test/tiering_tests.py#L198)):*
<details><summary><i>Previously, PatKamin (Patryk Kaminski) wrote…</i></summary><blockquote>

Comment out just the kind_name check.
</blockquote></details>

With this change if you pass FS_DAX kind name, the allocations will actually be made from it, so this is not only a name check.
___
*[tiering/memtier.c, line 119 at r1](https://reviewable.io/reviews/memkind/memkind/570#-MVqToU_B-wswN-Pryiv:-MVvMjSfE_vlyj62jtZf:b-896fix) ([raw file](https://github.com/memkind/memkind/blob/51a3ea47cc053d53fdd3ffd5ecd383d7b26ac526/tiering/memtier.c#L119)):*
<details><summary><i>Previously, michalbiesek wrote…</i></summary><blockquote>

IMHO: This should be a place to handle kind creation/ memkind_t setup, ctl module should be responsible for parsing string.
</blockquote></details>

Done.

**Review by michalbiesek — COMMENTED** [review]
> Reviewed 2 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/570)* status: 0 of 2 files reviewed, 5 unresolved discussions (waiting on @bratpiorka and @PatKamin)

___
*[test/tiering_tests.py, line 46 at r2](https://reviewable.io/reviews/memkind/memkind/570#-MVv5DKrChRgjuw0tFQj:-MVv5DKrChRgjuw0tFQk:bxq88nt) ([raw file](https://github.com/memkind/memkind/blob/6f95e499ae3997e2256be0ce40733a7c0492a7f4/test/tiering_tests.py#L46)):*
> ```
>         if kind_env == "DRAM":
>             return "memkind_default"
>         return ""
> ```
nit make this as dict, so you avoid in the future if-else ladder.
something like
> ```
> kind_name_dict= {'DRAM': 'memkind_default'}
> ```
and you will have only one return:
> ```
>  return kind_name_dict.get(kind_env, '')
> ```
and in case of extending this, you could just add a new key, value into dict
___
*[tiering/memtier.c, line 119 at r1](https://reviewable.io/reviews/memkind/memkind/570#-MVqToU_B-wswN-Pryiv:-MVv6bubE3fpUvVt5OG1:bns059c) ([raw file](https://github.com/memkind/memkind/blob/51a3ea47cc053d53fdd3ffd5ecd383d7b26ac526/tiering/memtier.c#L119)):*
Let me rephrase, since the current code obfuscates it.

This part of the code could be changed (moved the whole logic of create_tiered_kind_from_env(char *env_var_string) to ctl module)
See on this layer, you practically need only a tiered kind structure.

**current_tier** is only needed here for reference to freeing memory.


The ctl API should be modified in this manner:
- the ctl_load_config should be probably renamed and changed to the method, like
> ```
>struct memtier_tier ctl_create_tier_kind_from_env(char *env_var_string);
> ```

And extend for the destructing API method:

> ```
>void ctl_destroy_kind(struct memtier_tier tier);
> ```

Anyway, this could be done with:
https://github.com/memkind/memkind/blob/fd3352ccf3d4a78aecccd5f7b03a20d9fb6bbcf0/tiering/ctl.c#L146

I am not sure if and what should be done with this PR, but the comment above applies to solution with multiple kinds support
___
*[tiering/memtier.c, line 104 at r2](https://reviewable.io/reviews/memkind/memkind/570#-MVwobIpEAVU7b9Wy9Dt:-MVwobIpEAVU7b9Wy9Du:b-miatz2) ([raw file](https://github.com/memkind/memkind/blob/6f95e499ae3997e2256be0ce40733a7c0492a7f4/tiering/memtier.c#L104)):*
> ```
> map_str_to_kind
> ```
rename to get_kind
___
*[tiering/memtier.c, line 129 at r2](https://reviewable.io/reviews/memkind/memkind/570#-MVv6R6Y0Nw1Lv6KOBV2:-MVv6R6ZAtp8wtgsD-mO:b-cgwyv) ([raw file](https://github.com/memkind/memkind/blob/6f95e499ae3997e2256be0ce40733a7c0492a7f4/tiering/memtier.c#L129)):*
> ```
>     if (strcmp(kind_name, "FS_DAX") == 0) {
>         // TODO handle FS_DAX here
>         return -1;
> ```
why cannot this be in map_str_to_kind ? you could even move this log_debug there
___
*[tiering/memtier.c, line 132 at r2](https://reviewable.io/reviews/memkind/memkind/570#-MVv6sMgAPMoeSV40njJ:-MVv6sMgAPMoeSV40njK:bcxo0yn) ([raw file](https://github.com/memkind/memkind/blob/6f95e499ae3997e2256be0ce40733a7c0492a7f4/tiering/memtier.c#L132)):*
> ```C
>         // TODO handle FS_DAX here
>         return -1;
>     } else if (!(kind = map_str_to_kind(kind_name))) {
> ```

Change commit message to a real state of this PR:
- you provide parsing/mapping only for DRAM
- tieried kind -> tiered kind

**Review by PatKamin — COMMENTED** [review]
> Reviewed 4 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/570)* status: all files reviewed, 4 unresolved discussions (waiting on @bratpiorka)

**Review by bratpiorka — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/570)* status: 0 of 2 files reviewed, 3 unresolved discussions (waiting on @michalbiesek and @PatKamin)

___
*[test/tiering_tests.py, line 46 at r2](https://reviewable.io/reviews/memkind/memkind/570#-MVv5DKrChRgjuw0tFQj:-MVzQfNO1a53DLG24-DN:b3asexz) ([raw file](https://github.com/memkind/memkind/blob/6f95e499ae3997e2256be0ce40733a7c0492a7f4/test/tiering_tests.py#L46)):*
<details><summary><i>Previously, michalbiesek wrote…</i></summary><blockquote>

> ```
>         if kind_env == "DRAM":
>             return "memkind_default"
>         return ""
> ```
nit make this as dict, so you avoid in the future if-else ladder.
something like
> ```
> kind_name_dict= {'DRAM': 'memkind_default'}
> ```
and you will have only one return:
> ```
>  return kind_name_dict.get(kind_env, '')
> ```
and in case of extending this, you could just add a new key, value into dict
</blockquote></details>

I added this dict to Helper class
___
*[tiering/memtier.c, line 119 at r1](https://reviewable.io/reviews/memkind/memkind/570#-MVqToU_B-wswN-Pryiv:-MVzRJq52hicbfOU2xCj:b-nioxiw) ([raw file](https://github.com/memkind/memkind/blob/51a3ea47cc053d53fdd3ffd5ecd383d7b26ac526/tiering/memtier.c#L119)):*
<details><summary><i>Previously, michalbiesek wrote…</i></summary><blockquote>

Let me rephrase, since the current code obfuscates it.

This part of the code could be changed (moved the whole logic of create_tiered_kind_from_env(char *env_var_string) to ctl module)
See on this layer, you practically need only a tiered kind structure.

**current_tier** is only needed here for reference to freeing memory.


The ctl API should be modified in this manner:
- the ctl_load_config should be probably renamed and changed to the method, like
> ```
>struct memtier_tier ctl_create_tier_kind_from_env(char *env_var_string);
> ```

And extend for the destructing API method:

> ```
>void ctl_destroy_kind(struct memtier_tier tier);
> ```

Anyway, this could be done with:
https://github.com/memkind/memkind/blob/fd3352ccf3d4a78aecccd5f7b03a20d9fb6bbcf0/tiering/ctl.c#L146

I am not sure if and what should be done with this PR, but the comment above applies to solution with multiple kinds support

</blockquote></details>

I fully agree that this part of code should be refactored. But the API you proposed should return a list of tiered kinds (not only one) and because currently few peoples works on this part of code I don't want to change it much. Let's do this refactoring after (or as a part of) handling multiple kinds in env var. I added a TODO before create_tiered_kind_from_env() definition about this.
___
*[tiering/memtier.c, line 104 at r2](https://reviewable.io/reviews/memkind/memkind/570#-MVwobIpEAVU7b9Wy9Dt:-MVzQkatCeedmR4oCJrE:b-896fix) ([raw file](https://github.com/memkind/memkind/blob/6f95e499ae3997e2256be0ce40733a7c0492a7f4/tiering/memtier.c#L104)):*
<details><summary><i>Previously, michalbiesek wrote…</i></summary><blockquote>

> ```
> map_str_to_kind
> ```
rename to get_kind
</blockquote></details>

Done.
___
*[tiering/memtier.c, line 129 at r2](https://reviewable.io/reviews/memkind/memkind/570#-MVv6R6Y0Nw1Lv6KOBV2:-MVzRCuz99W-ND2Z0zqx:b-896fix) ([raw file](https://github.com/memkind/memkind/blob/6f95e499ae3997e2256be0ce40733a7c0492a7f4/tiering/memtier.c#L129)):*
<details><summary><i>Previously, michalbiesek wrote…</i></summary><blockquote>

> ```
>     if (strcmp(kind_name, "FS_DAX") == 0) {
>         // TODO handle FS_DAX here
>         return -1;
> ```
why cannot this be in map_str_to_kind ? you could even move this log_debug there
</blockquote></details>

Done.
___
*[tiering/memtier.c, line 132 at r2](https://reviewable.io/reviews/memkind/memkind/570#-MVv6sMgAPMoeSV40njJ:-MVzRDYU62rWYq0OIIqL:bb6un1j) ([raw file](https://github.com/memkind/memkind/blob/6f95e499ae3997e2256be0ce40733a7c0492a7f4/tiering/memtier.c#L132)):*
<details><summary><i>Previously, michalbiesek wrote…</i></summary><blockquote>

Change commit message to a real state of this PR:
- you provide parsing/mapping only for DRAM
- tieried kind -> tiered kind

</blockquote></details>

Done

**Review by michalbiesek — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/570)* status: 0 of 2 files reviewed, all discussions resolved (waiting on @PatKamin)

**Review by PatKamin — COMMENTED** [review]
> Reviewed 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/570)* status: all files reviewed, 1 unresolved discussion (waiting on @bratpiorka)

___
*[tiering/memtier.c, line 105 at r3](https://reviewable.io/reviews/memkind/memkind/570#-MVzlp5g7Tv0ZfAIv1H-:-MVzlp5g7Tv0ZfAIv1H0:br2xuwz) ([raw file](https://github.com/memkind/memkind/blob/8479ab393457686967f67a63b8ef56c82d87d947/tiering/memtier.c#L105)):*
> ```C
> 
> static memkind_t get_kind(const char *str, const char *pmem_path,
>                           const char *pmem_size)
> ```

This function should only do this: get_kind. Remove pmem_path and pmem_size.

**Review by michalbiesek — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/570)* status: all files reviewed, 1 unresolved discussion (waiting on @PatKamin)

___
*[tiering/memtier.c, line 105 at r3](https://reviewable.io/reviews/memkind/memkind/570#-MVzlp5g7Tv0ZfAIv1H-:-MVzmS01B2wAnzyX0PPt:b-jjs0i) ([raw file](https://github.com/memkind/memkind/blob/8479ab393457686967f67a63b8ef56c82d87d947/tiering/memtier.c#L105)):*
<details><summary><i>Previously, PatKamin (Patryk Kaminski) wrote…</i></summary><blockquote>

This function should only do this: get_kind. Remove pmem_path and pmem_size.
</blockquote></details>

IMHO it could be done as it is. See that you have all that is required here to create FS-DAX kind and return it as memkind_t object, besides this probably be later moved to ctl module so no need to change it here.

**Review by PatKamin — COMMENTED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/570#-:-MVzvdIW2cwMepoBpvrD:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @bratpiorka)
